Add rollout option to osconfig patch deployment fixes https://github.com/hashicorp/terraform-provider-google/issues/7124

cookejames opened the pull request:
<!-- Put a description of what this PR is for here, along with any references to issues that this resolves or contributes to -->




<!--
Replace each [ ] with [X] to check it. Switch to the preview view to make it easier to click on links.
These steps will speed up the review process, and we appreciate you spending time on them before sending
your code to be reviewed.
-->
If this PR is for Terraform, I acknowledge that I have:

- [x] Searched through the [issue tracker](https://github.com/hashicorp/terraform-provider-google/issues) for an open issue that this either resolves or contributes to, commented on it to claim it, and written "fixes {url}" or "part of {url}" in this PR description. If there were no relevant open issues, I opened one and commented that I would like to work on it (not necessary for very small changes).
- [x] [Generated Terraform](https://github.com/GoogleCloudPlatform/magic-modules#generating-downstream-tools), and ran `make test` and `make lint` to ensure it passes unit and linter tests.
- [x] Ensured that all new fields I added that can be set by a user appear in at least one [example](https://github.com/GoogleCloudPlatform/magic-modules/tree/master/templates/terraform/examples) (for generated resources) or [third_party test](https://github.com/GoogleCloudPlatform/magic-modules/tree/master/third_party/terraform/tests) (for handwritten resources or update tests).
- [ ] [Ran](https://github.com/hashicorp/terraform-provider-google/blob/master/.github/CONTRIBUTING.md#tests) relevant acceptance tests (If the acceptance tests do not yet pass or you are unable to run them, please let your reviewer know).
- [x] Read the [Release Notes Guide](https://github.com/GoogleCloudPlatform/magic-modules/blob/master/.ci/RELEASE_NOTES_GUIDE.md) before writing my release note below.

I have not tried to run the [contributing tests](https://github.com/hashicorp/terraform-provider-google/blob/master/.github/CONTRIBUTING.md#tests) yet. `make test`, and compiling the terraform module work.

`make lint` seems to run correctly for the osconfig module but has warning for other existing modules. These errors exist in master.

<!-- AUTOCHANGELOG for Downstream PRs.

Please select one of the following "release-note:" headings:
    - release-note:enhancement
    - release-note:bug
    - release-note:note
    - release-note:new-resource
    - release-note:new-datasource
    - release-note:deprecation
    - release-note:breaking-change
    - release-note:none
    
Unless you choose release-note:none, please add a release note.

See .ci/RELEASE_NOTES_GUIDE.md for writing good release notes.

You can add more release note blocks if you want more than one CHANGELOG
entry for this PR.
-->
**Release Note Template for Downstream PRs (will be copied)**

```release-note:enhancement
osconfig: added rollout field to `google_os_config_patch_deployment`
```

modular-magician commented:
Hello!  I am a robot who works on Magic Modules PRs.

I have detected that you are a community contributor, so your PR will be assigned to someone with a commit-bit on this repo for initial review.

Thanks for your contribution!  A human will be with you soon.

@slevenick, please review this PR or find an appropriate assignee.

cookejames commented:
@googlebot I signed it!

modular-magician commented:
Hi!  I'm the modular magician.  Your PR generated some diffs in downstreams - here they are.
# Diff report:


Terraform GA: [Diff](https://github.com/modular-magician/terraform-provider-google/compare/auto-pr-3919-old..auto-pr-3919) ( 2 files changed, 222 insertions(+))
Terraform Beta: [Diff](https://github.com/modular-magician/terraform-provider-google-beta/compare/auto-pr-3919-old..auto-pr-3919) ( 2 files changed, 222 insertions(+))
TF Conversion: [Diff](https://github.com/modular-magician/terraform-google-conversion/compare/auto-pr-3919-old..auto-pr-3919) ( 1 file changed, 70 insertions(+))

slevenick commented:
Hey @cookejames thanks for the contribution!

Can you add a test for this feature? It should be as easy as adding the rollout block to this file: https://github.com/GoogleCloudPlatform/magic-modules/blob/master/templates/terraform/examples/os_config_patch_deployment_full.tf.erb

which gets generated into the test for this resource.

slevenick commented on products/osconfig/api.yaml:
```diff
@@ -844,6 +844,41 @@ objects:
+          Rollout strategy of the patch job.
+        properties:
+          - !ruby/object:Api::Type::Enum
+            name: 'mode'
```
I'd guess this should be required? Any reason a user would not want to specify the mode if they are specifying rollout?

cookejames replied:
Yes it should

slevenick commented on products/osconfig/api.yaml:
```diff
@@ -844,6 +844,41 @@ objects:
+              - :ZONE_BY_ZONE
+              - :CONCURRENT_ZONES
+          - !ruby/object:Api::Type::NestedObject
+            name: 'disruptionBudget'
```
Same as above, should this be required?

slevenick commented on products/osconfig/api.yaml:
```diff
@@ -844,6 +844,41 @@ objects:
+                description: |
+                  Specifies a fixed value.
+                exactly_one_of:
+                  - fixed
```
I believe this should be `rollout.0.disruption_budget.0.fixed` and `rollout.0.disruption_budget.0.percentage`

We have to specify the full path as these get evaluated from the top-level

cookejames replied:
Thanks

cookejames commented:
Thanks for the comments this is the first time I've worked on magic modules so didn't realise that was an automated test I thought it was more for generating documentation. I've added a test and fixed the other points you have mentioned.

modular-magician commented:
Hi!  I'm the modular magician.  Your PR generated some diffs in downstreams - here they are.
# Diff report:


Terraform GA: [Diff](https://github.com/modular-magician/terraform-provider-google/compare/auto-pr-3919-old..auto-pr-3919) ( 3 files changed, 236 insertions(+))
Terraform Beta: [Diff](https://github.com/modular-magician/terraform-provider-google-beta/compare/auto-pr-3919-old..auto-pr-3919) ( 3 files changed, 236 insertions(+))
TF Conversion: [Diff](https://github.com/modular-magician/terraform-google-conversion/compare/auto-pr-3919-old..auto-pr-3919) ( 1 file changed, 70 insertions(+))
TF OiCS: [Diff](https://github.com/modular-magician/docs-examples/compare/auto-pr-3919-old..auto-pr-3919) ( 1 file changed, 7 insertions(+))

slevenick commented on templates/terraform/examples/os_config_patch_deployment_full.tf.erb:
```diff
@@ -94,4 +94,11 @@ resource "google_os_config_patch_deployment" "<%= ctx[:primary_resource_id] %>"
+
+  rollout {
+    mode = "ZONE_BY_ZONE"
+    disruptionBudget {
```
Looks like this should be `disruption_budget`

modular-magician commented:
Hi!  I'm the modular magician.  Your PR generated some diffs in downstreams - here they are.
# Diff report:


Terraform GA: [Diff](https://github.com/modular-magician/terraform-provider-google/compare/auto-pr-3919-old..auto-pr-3919) ( 3 files changed, 236 insertions(+))
Terraform Beta: [Diff](https://github.com/modular-magician/terraform-provider-google-beta/compare/auto-pr-3919-old..auto-pr-3919) ( 3 files changed, 236 insertions(+))
TF Conversion: [Diff](https://github.com/modular-magician/terraform-google-conversion/compare/auto-pr-3919-old..auto-pr-3919) ( 1 file changed, 70 insertions(+))
TF OiCS: [Diff](https://github.com/modular-magician/docs-examples/compare/auto-pr-3919-old..auto-pr-3919) ( 1 file changed, 7 insertions(+))
